File: services/detect_fraud/src/consumers/transactionConsumer.js
const { Kafka, CompressionTypes } = require('kafkajs');
const config = require('../config');
const logger = require('../config/logger');
const fraudDetectionService = require('../services/fraudDetectionService');
const decisionPublisher = require('../services/decisionPublisher');

let consumer = null;
let producer = null;

const normalizeTransaction = (payload) => {
  const raw = payload?.transaction || payload?.originalTransaction || payload?.data || payload;
  const metadata = raw?.metadata || {};
  const createdAt = raw?.createdAt || payload?.createdAt || new Date().toISOString();

  return {
    id: raw?.id || raw?.transactionId || raw?.transaction_id,
    customerId: raw?.customerId || raw?.customer_id || payload?.customerId || payload?.customer_id,
    merchantId: raw?.merchantId || raw?.merchant_id || payload?.merchantId || payload?.merchant_id || 'FTDS_TRANSFER',
    amount: Number(raw?.amount),
    currency: raw?.currency || 'SGD',
    cardType: raw?.cardType || raw?.card_type || 'CREDIT',
    createdAt,
    location: raw?.location || { country: raw?.country || 'SG' },
    metadata,
  };
};

const validateTransaction = (transaction) => {
  if (!transaction.id) {
    return 'transaction.id is required';
  }
  if (!transaction.customerId) {
    return 'transaction.customerId is required';
  }
  if (!Number.isFinite(transaction.amount)) {
    return 'transaction.amount must be a finite number';
  }
  return null;
};

const commitOffset = async (topic, partition, offset) => {
  if (!consumer) return;
  await consumer.commitOffsets([
    {
      topic,
      partition,
      offset: (BigInt(offset) + 1n).toString(),
    },
  ]);
};

const sendToDlq = async ({ topic, partition, offset, reason, raw = null, payload = null, error = null }) => {
  if (!producer) {
    throw new Error('Detect fraud DLQ producer is not ready');
  }

  await producer.send({
    topic: config.kafka.dlqTopic,
    compression: CompressionTypes.GZIP,
    messages: [
      {
        key: String(payload?.transactionId || payload?.transaction?.id || topic),
        value: JSON.stringify({
          eventType: 'detect-fraud.dlq',
          sourceTopic: topic,
          sourcePartition: partition,
          sourceOffset: offset,
          reason,
          error,
          rawPayload: raw,
          originalPayload: payload,
          failedAt: new Date().toISOString(),
          serviceName: config.serviceName,
        }),
        headers: {
          'content-type': 'application/json',
          'service-source': config.serviceName,
          'x-dlq-reason': reason,
        },
      },
    ],
  });
};

const start = async () => {
  if (consumer) return;

  const kafka = new Kafka({
    clientId: config.kafka.clientId,
    brokers: config.kafka.brokers,
    retry: config.kafka.retry,
  });

  consumer = kafka.consumer({
    groupId: config.kafka.groupId,
    allowAutoTopicCreation: false,
    retry: config.kafka.retry,
  });

  producer = kafka.producer({
    allowAutoTopicCreation: false,
    idempotent: true,
    maxInFlightRequests: 1,
    retry: config.kafka.retry,
  });

  await consumer.connect();
  await producer.connect();
  await consumer.subscribe({ topic: config.kafka.inputTopic, fromBeginning: false });

  await consumer.run({
    autoCommit: false,
    eachMessage: async ({ topic, partition, message, heartbeat }) => {
      const offset = message.offset;
      const raw = message.value?.toString();

      if (!raw) {
        logger.warn('Sending empty transaction.created event to DLQ', { topic, partition, offset });
        await sendToDlq({ topic, partition, offset, reason: 'empty_payload' });
        await commitOffset(topic, partition, offset);
        return;
      }

      let payload;
      try {
        payload = JSON.parse(raw);
      } catch (error) {
        logger.error('Sending malformed transaction.created event to DLQ', {
          topic,
          partition,
          offset,
          error: error.message,
        });
        await sendToDlq({
          topic,
          partition,
          offset,
          reason: 'parse_error',
          raw,
          error: error.message,
        });
        await commitOffset(topic, partition, offset);
        return;
      }

      try {
        const transaction = normalizeTransaction(payload);
        const validationError = validateTransaction(transaction);
        if (validationError) {
          logger.error('Sending invalid transaction.created event to DLQ', {
            topic,
            partition,
            offset,
            error: validationError,
          });
          await sendToDlq({
            topic,
            partition,
            offset,
            reason: 'invalid_event',
            raw,
            payload,
            error: validationError,
          });
          await commitOffset(topic, partition, offset);
          return;
        }

        await heartbeat();
        const fraudAnalysis = await fraudDetectionService.analyzeTransaction(transaction);
        await heartbeat();

        await producer.send({
          topic: config.kafka.outputTopic,
          compression: CompressionTypes.GZIP,
          messages: [
            {
              key: String(transaction.customerId),
              value: JSON.stringify({
                eventType: 'transaction.scored',
                transactionId: transaction.id,
                customerId: transaction.customerId,
                merchantId: transaction.merchantId,
                correlationId: payload.correlationId || payload.trace_id || transaction.id,
                originalTransaction: transaction,
                fraudAnalysis,
                data: {
                  transaction_id: transaction.id,
                  rules_score: fraudAnalysis.riskScore,
                  reason: fraudAnalysis.reasons[0] || null,
                },
                processedAt: new Date().toISOString(),
              }),
              headers: {
                'content-type': 'application/json',
                'service-source': config.serviceName,
              },
            },
          ],
        });

        await heartbeat();
        await decisionPublisher.process({
          producer,
          transaction,
          fraudAnalysis,
          correlationId: payload.correlationId || payload.trace_id || transaction.id,
        });

        await commitOffset(topic, partition, offset);
      } catch (error) {
        logger.error('Fraud detection consumer failed to process transaction', {
          topic,
          partition,
          offset,
          error: error.message,
        });
        throw error;
      }
    },
  });
};

const stop = async () => {
  if (consumer) {
    await consumer.disconnect();
    consumer = null;
  }
  if (producer) {
    await producer.disconnect();
    producer = null;
  }
};

module.exports = {
  start,
  stop,
};

File: services/detect_fraud/src/consumers/transaction_consumer.py
from __future__ import annotations

import logging
import math
from datetime import datetime, timezone
from typing import Any

from services.detect_fraud.src.config.settings import settings

logger = logging.getLogger("detect_fraud")


def _header(name: str, value: str | None) -> tuple[str, bytes]:
    return (name, (value or "").encode("utf-8"))


def normalize_transaction(payload: dict[str, Any]) -> dict[str, Any]:
    raw = payload.get("transaction") or payload.get("originalTransaction") or payload.get("data") or payload
    metadata = raw.get("metadata") if isinstance(raw.get("metadata"), dict) else {}
    created_at = raw.get("createdAt") or payload.get("createdAt") or datetime.now(timezone.utc).isoformat()

    return {
        "id": raw.get("id") or raw.get("transactionId") or raw.get("transaction_id"),
        "customerId": (
            raw.get("customerId") or raw.get("customer_id")
            or payload.get("customerId") or payload.get("customer_id")
        ),
        "merchantId": (
            raw.get("merchantId") or raw.get("merchant_id")
            or payload.get("merchantId") or payload.get("merchant_id")
            or "FTDS_TRANSFER"
        ),
        "amount": float(raw.get("amount")) if raw.get("amount") is not None else float("nan"),
        "currency": raw.get("currency") or "SGD",
        "cardType": raw.get("cardType") or raw.get("card_type") or "CREDIT",
        "createdAt": created_at,
        "location": raw.get("location") or {"country": raw.get("country") or "SG"},
        "metadata": metadata,
    }


def validate_transaction(transaction: dict[str, Any]) -> str | None:
    if not transaction.get("id"):
        return "transaction.id is required"
    if not transaction.get("customerId"):
        return "transaction.customerId is required"
    amount = transaction.get("amount")
    if not isinstance(amount, (int, float)) or not math.isfinite(float(amount)):
        return "transaction.amount must be a finite number"
    return None


async def send_to_dlq(
    producer: Any,
    *,
    topic: str,
    partition: int,
    offset: int,
    reason: str,
    raw_payload: str | None = None,
    parsed_payload: dict[str, Any] | None = None,
    error: str | None = None,
) -> None:
    if producer is None:
        raise RuntimeError("Detect fraud DLQ producer is not ready")

    key = (
        parsed_payload.get("transactionId") if parsed_payload else None
    ) or (
        parsed_payload.get("transaction", {}).get("id")
        if parsed_payload and isinstance(parsed_payload.get("transaction"), dict)
        else None
    ) or topic

    await producer.send_and_wait(
        settings.kafka_dlq_topic,
        key=str(key).encode("utf-8"),
        value={
            "eventType": "detect-fraud.dlq",
            "sourceTopic": topic,
            "sourcePartition": partition,
            "sourceOffset": offset,
            "reason": reason,
            "error": error,
            "rawPayload": raw_payload,
            "originalPayload": parsed_payload,
            "failedAt": datetime.now(timezone.utc).isoformat(),
            "serviceName": settings.service_name,
        },
        headers=[
            _header("content-type", "application/json"),
            _header("service-source", settings.service_name),
            _header("x-dlq-reason", reason),
        ],
    )
